compiler/testData/diagnostics/tests/enum/incompatibleEnums_withExtendedTypeCompatibilityChecker.fir.kt

```diff
@@ -0,0 +1,145 @@
+// !LANGUAGE: +ProhibitComparisonOfIncompatibleEnums +ExtendedTypeCompatibilityChecker
+
+enum class E1 {
+    A, B
+}
+
+enum class E2 {
+    A, B
+}
+
+fun foo1(e1: E1, e2: E2) {
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 != e2<!>
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.B == e2<!>
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.A == E2.B<!>
+
+    e1 == E1.A
+    E1.A == e1
+    e2 == E2.B
+    E2.B == e2
+}
+
+fun foo2(e1: E1, e2: E2) {
+    when (e1) {
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.B<!> -> {}
+        e1 -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        else -> {}
+    }
+}
+
+fun foo3(e1: Enum<E1>, e2: Enum<E2>, e: Enum<*>) {
+    e1 == e
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+
+    e1 == E1.A
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+
+    when (e1) {
+        e1 -> {}
+        <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        e -> {}
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+        else -> {}
+    }
+
+    when (e) {
+        e -> {}
+        e2 -> {}
+        E1.A -> {}
+        E2.A -> {}
+        else -> {}
+    }
+}
+
+interface MyInterface
+open class MyOpenClass
+
+fun foo4(e1: E1, i: MyInterface, c: MyOpenClass) {
+    <!EQUALITY_NOT_APPLICABLE!>e1 == i<!>
+    <!EQUALITY_NOT_APPLICABLE!>i == e1<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>e1 == c<!>
+    <!EQUALITY_NOT_APPLICABLE!>c == e1<!>
+
+    when (e1) {
+            <!INCOMPATIBLE_TYPES!>i<!> -> {}
+            <!INCOMPATIBLE_TYPES!>c<!> -> {}
+        else -> {}
+    }
+}
+
+enum class E3 : MyInterface { X, Y }
+
+fun foo5(i: MyInterface, a: Any) {
+    E3.X == E3.Y
+    E3.X == i
+    E3.X == a
+}
+
+fun foo6(e1: E1?, e2: E2) {
+    E1.A == null
+    null == E1.A
```

Comment:
Just curious, is it easy/trivial to add `SENSELESS_COMPARISON` here?

Reply:
I think we can cover some cases of `SENSELESS_COMPARISON` fairly easily. But let me do a follow up PR that covers it more fully.